internal/dag/dag.go

```diff
@@ -71,6 +71,56 @@ func (g *AcyclicGraph) Descendants(v Vertex) Set {
 	return s
 }
 
+// FirstDescendantsWith returns a Set that includes every Vertex yielded by
+// walking up from the provided starting Vertex v, and stopping each branch when
+// match returns true. This will return the set of all first descendants
+// encountered which match some criteria.
+func (g *AcyclicGraph) FirstDescendantsWith(v Vertex, match func(Vertex) bool) Set {
+	s := make(Set)
+	searchFunc := func(v Vertex, d int) error {
+		if match(v) {
+			s.Add(v)
+			return errStopWalkBranch
```

Comment:
Dumb question, but why are errors used as the signal for the walker on what to do? I was a bit confused when the comment said doesn't return an error but the function signature said it does :D  Couldn't this be any other (string based / iota based) enum type?

Reply:
The sentinel error values were used so that I didn't need to refactor the walk methods entirely, and only needed a simple switch on the error value. All 4 walk methods still do return the error from the callback since you can't use the stop sentinels externally, and could still be called in the same way as before. It's only the more specialized wrappers like `Descendants` which didn't use an error at all that make use of these special sentinels error values.

If we wanted to generalize this as a shared library, then we would probably want to rewrite the walks to use an enum+error return for some use case that we don't need in Terraform ;)